app/routes/user/endorsements.server.ts

```diff
@@ -0,0 +1,214 @@
+/*!
+ * Copyright © 2022 United States Government as represented by the Administrator
+ * of the National Aeronautics and Space Administration. No copyright is claimed
+ * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+ *
+ * SPDX-License-Identifier: NASA-1.3
+ */
+
+import { tables } from '@architect/functions'
+import { getSubmitterUsersFromCognito } from '~/lib/utils'
+import { getUser } from '../__auth/user.server'
+
+// models
+export type EndorsementRequest = {
+  endorserEmail: string
+  endorserSub: string
+  requestorEmail: string
+  requestorSub: string
+  created: number
+  status: EndorsementStates
+}
+
+export type EndorsementStates = 'approved' | 'pending' | 'rejected' | 'reported'
+
+export class EndorsementsServer {
+  #sub: string
+
+  private constructor(sub: string) {
+    this.#sub = sub
+  }
+
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user)
+      throw new Response('not signed in', {
+        statusText: 'not signed in',
+        status: 403,
+      })
+    return new this(user.sub)
+  }
+
+  async createEndorsementRequest(request: EndorsementRequest) {
+    // Checking endorser sub:
+    if (!this.requestedEndorserIsValid(request.endorserSub))
+      throw new Response('Requested Endorser is not a verified submitter', {
+        statusText: 'Bad Request',
+        status: 400,
+      })
+    const db = await tables()
+    await db.circular_endorsements.put(request)
+  }
+
+  // Allows the current user to change the requested endorser on their previously made request
+  async updateRequestedEndorser(
+    existingEndorserSub: string,
+    newEndorserSub: string,
+    newEndorserEmail: string
+  ) {
+    const db = await tables()
+
+    const existingEndorsement = (await db.circular_endorsements.get({
+      requestorSub: this.#sub,
+      endorserSub: existingEndorserSub,
+    })) as EndorsementRequest
+
+    this.validateStateAndExistanceOfRequest(existingEndorsement)
+
+    await db.circular_endorsements.update({
+      Key: {
+        requestorSub: this.#sub,
+        endorserSub: existingEndorserSub,
+      },
+      UpdateExpression:
+        'set #endorserEmail = :newEndorserEmail, #endorserSub = :newEndorserSub',
+      ExpressionAttributeNames: {
+        '#endorserSub': 'endorserSub',
+        '#endorserEmail': 'endorserEmail',
+      },
+      ExpressionAttributeValues: {
+        ':newEndorserSub': newEndorserSub,
+        ':newEndorserEmail': newEndorserEmail,
+      },
+    })
+  }
+
+  // Allows the current user to update the status of a request they have been listed on as an endorser
+  async updateEndorsementRequestStatus(
+    status: EndorsementStates,
+    requestorSub: string
+  ) {
+    if (!this.requestedEndorserIsValid(this.#sub))
```

Comment:
Please check if the user is authorized by looking at the group claims in their access token.